src/osyris/spatial/spatial.py

```diff
@@ -0,0 +1,147 @@
+
+
+def rotation_matrix(vec, angle):
+    """
+    Returns 3D rotation matrix of angle 'angle' around rotation vector 'vec'.
+    """
+    if isinstance(vec, list):
+        vec = np.array(vec)
+    vec = vec / np.linalg.norm(vec)
+    R = np.cos(angle) * np.identity(3) + (np.sin(angle)) * np.cross(
+        vec,
+        np.identity(vec.shape[0]) * -1) + (1 - np.cos(angle)) * (np.outer(vec, vec))
+    return R
+
+
+def cartesian_to_spherical_rotation_matrix(colatitude, azimuth):
+    """
+    Used to convert arbitrary cartesian vectors into spherical vectors
+    """
+
+    R = np.array([[
+        np.sin(colatitude) * np.cos(azimuth),
+        np.sin(colatitude) * np.sin(azimuth),
+        np.cos(colatitude)
+    ],
+                  [
+                      np.cos(colatitude) * np.cos(azimuth),
+                      np.cos(colatitude) * np.sin(azimuth), -np.sin(colatitude)
+                  ], [-np.sin(azimuth), np.cos(azimuth), 0]],
+                 dtype=object)
+    return R
+
+
+def get_ang_mom(subdomain, dr_L):
+    """
+    Compute angular momentum vector in sphere of radius dr_L
+    """
+    sphere = (subdomain["amr"]["position"].norm <= dr_L).values
+    pos = subdomain["amr"]["position"][sphere]
+    mv = subdomain["hydro"]["mass"][sphere] * subdomain["hydro"]["velocity"][sphere]
+    L = np.sum(pos.cross(mv))
+    return L
+
+
+def change_basis(subdomain, new_basis):
+    """
+    Rotates all vectors in dataset to align with vector in 'new_basis'
+    """
+    try:
+        old_basis = subdomain.basis
+    except AttributeError:
+        old_basis = [0, 0, 1]  # assume it's the ramses grid
+    if hasattr(new_basis, 'nvec'):  # if it's a vector
+        new_basis = [new_basis.x.values, new_basis.y.values, new_basis.z.values]
+    rot_angle = np.arccos(
+        np.dot(old_basis, new_basis) /
+        (np.linalg.norm(old_basis) * np.linalg.norm(new_basis)))
+    rot_vector = np.cross(new_basis, old_basis)
+    R = rotation_matrix(rot_vector, rot_angle)
+    for g in subdomain.groups.keys():
+        for element in subdomain[g]:
+            if hasattr(subdomain[g][element], 'nvec'):
+                # all of this will be simplified once matmul is integraded into Vector
+                vector = np.array([
+                    subdomain[g][element].x.values, subdomain[g][element].y.values,
+                    subdomain[g][element].z.values
+                ])
+                vector = R @ vector
+                subdomain[g][element] = subdomain[g][element].__class__(
+                    x=vector[0],
+                    y=vector[1],
+                    z=vector[2],
+                    unit=subdomain[g][element].unit)
+    subdomain.basis = new_basis
+
+
+def get_spherical_colatitude(position_vec):
+    """
+    Compute colatitude from cartesian position vector
+    """
+    X = position_vec.x.values
+    Y = position_vec.y.values
+    Z = position_vec.z.values
+
+    theta = np.arctan2(np.linalg.norm([X, Y], axis=0), Z)
```

Comment:
Can you do computations with Arrays directly instead of having to extract the values?
E.g. `np.arccos(position.z / position.r)` ?